sotodlib/site_pipeline/preprocess_tod.py

```diff
@@ -371,34 +371,67 @@ def _main(executor: Union["MPICommExecutor", "ProcessPoolExecutor"],
     n_fail = 0
 
     # Run write_block obs-ids in parallel at once then write all to the sqlite db.
-    futures = [executor.submit(preprocess_tod, obs_id=r[0]['obs_id'],
-                    group_list=r[1], verbosity=verbosity,
-                    configs=configs,
-                    overwrite=overwrite, run_parallel=True) for r in run_list]
-    for future in as_completed_callable(futures):
-        logger.info('New future as_completed result')
-        try:
-            err, db_datasets = future.result()
+    futures = []
+    futures_kargs = {}
+    for r in run_list:
+        futures_kargs[r[0]["obs_id"]] = {
+            "obs_id": r[0]["obs_id"],
+            "group_list": r[1],
+            "verbosity": verbosity,
+            "configs": configs,
+            "overwrite": overwrite,
+            "run_parallel": True,
+            "try": 1,
+        }
+        kwargs = {k: v for k, v in futures_kargs[r[0]["obs_id"]].items() if k != "try"}
+        fut = executor.submit(
+            preprocess_tod,
+            **kwargs
+        )
+        fut.obs_id = r[0]["obs_id"]
+        futures.append(fut)
+
+    while futures:
+        future_to_check = futures.pop(0)
+        kwargs = {k: v for k, v in futures_kargs[future_to_check.obs_id].items() if k != "try"}
+        if future_to_check.done() and future_to_check.exception() is None:
```

Comment:
The `if` is being extended a bit. First all cases. check if the future has finished running, and then it check the status of the exception. If everything went well the same as before, otherwise checks the type of exception. If it is an OSError it resubmits (`executor.submit`) it. If the OSError appeared more times than the limit it stops. If the error is not an OSError it stops it. When none of this is true it just puts it at the end of the list to check.